### PR TITLE
Added Tree traversal in Crystal-lang

### PR DESCRIPTION
**Pull request by NIFR91**
Added Tree traversal algorithm in Crystal-lang. 

The name of all functions are as in the available implementations, added a new function just to print 
the tree, so it is easier to see the traversal.  Should I remove it?. 

Currently the script's output is: 

```text
 Creating Tree
(2)
 ├(1)
 │ ├(0)
 │ ├(0)
 │ └(0)
 ├(1)
 │ ├(0)
 │ ├(0)
 │ └(0)
 └(1)
   ├(0)
   ├(0)
   └(0)
Using recursive DFS:
2100010001000
Using recursive DFS with post-order traversal:
0001000100012
Using stack-based DFS:
2100010001000
Using queue-based BFS:
2111000000000
Creating binary tree to test in-order traversal
(3)
 ├(2)
 │ ├(1)
 │ │ ├(0)
 │ │ └(0)
 │ └(1)
 │   ├(0)
 │   └(0)
 └(2)
   ├(1)
   │ ├(0)
   │ └(0)
   └(1)
     ├(0)
     └(0)
Using In-order DFS:
010201030102010
```

### CODE REVIEW

File: book.json [hunk 1]
```diff
@@ -128,6 +128,7 @@
           "lang": "php",
           "name": "PHP"
         }
+
```
Comment (jiegillet) [inline book.json]: This change is unnecessary

File: contents/tree_traversal/code/crystal/tree-traversal.cr [hunk 2]
```diff
@@ -0,0 +1,116 @@
+class Node 
+  property id,children
```
Comment (jiegillet) [inline contents/tree_traversal/code/crystal/tree-traversal.cr]: minor nitpick: add a space after the comma

File: contents/tree_traversal/code/crystal/tree-traversal.cr [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  queue = Deque.new [node]
+
+  until queue.empty? 
+    temp = queue.shift
+    print temp.id 
+    temp.children.each{ |child| queue.push child }
+  end  
+end 
+
+def create_tree(levels,num_childs)
```
Comment (jiegillet) [inline contents/tree_traversal/code/crystal/tree-traversal.cr]: same nitpick :p

File: contents/tree_traversal/code/crystal/tree-traversal.cr [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  until queue.empty? 
+    temp = queue.shift
+    print temp.id 
+    temp.children.each{ |child| queue.push child }
+  end  
+end 
+
+def create_tree(levels,num_childs) 
+
+  children = [ ] of Node
```
Comment (jiegillet) [inline contents/tree_traversal/code/crystal/tree-traversal.cr]: Do you need the space in `[ ]`?

File: contents/tree_traversal/code/crystal/tree-traversal.cr [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    print temp.id 
+    temp.children.each{ |child| queue.push child }
+  end  
+end 
+
+def create_tree(levels,num_childs) 
+
+  children = [ ] of Node  
+  unless levels == 0 
+    num_childs.times{children.push create_tree levels-1,num_childs }
```
Comment (jiegillet) [inline contents/tree_traversal/code/crystal/tree-traversal.cr]: space after comma

File: contents/tree_traversal/code/crystal/tree-traversal.cr [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+  children = [ ] of Node  
+  unless levels == 0 
+    num_childs.times{children.push create_tree levels-1,num_childs } 
+  end 
+  
+  Node.new(levels,children) 
+end
+
+def print_tree(node,depth = [] of String)
```
Comment (jiegillet) [inline contents/tree_traversal/code/crystal/tree-traversal.cr]: Did you write this function yourself? It's quite neat!

File: book.json (hunk 1 above)
Comment (NIFR91) [reply book.json]: Fixed when I merged the current master branch

File: contents/tree_traversal/code/crystal/tree-traversal.cr (hunk 2 above)
Comment (NIFR91) [reply contents/tree_traversal/code/crystal/tree-traversal.cr]: Done, added spaces after commas

File: contents/tree_traversal/code/crystal/tree-traversal.cr (hunk 3 above)
Comment (NIFR91) [reply contents/tree_traversal/code/crystal/tree-traversal.cr]: Idem

File: contents/tree_traversal/code/crystal/tree-traversal.cr (hunk 4 above)
Comment (NIFR91) [reply contents/tree_traversal/code/crystal/tree-traversal.cr]: It is not necessary

File: contents/tree_traversal/code/crystal/tree-traversal.cr (hunk 5 above)
Comment (NIFR91) [reply contents/tree_traversal/code/crystal/tree-traversal.cr]: Done

File: contents/tree_traversal/code/crystal/tree-traversal.cr (hunk 6 above)
Comment (NIFR91) [reply contents/tree_traversal/code/crystal/tree-traversal.cr]: I wrote it based on Steven Pigeon's Blog Post  ["Pretty Printing a Tree in a Terminal"](https://hbfs.wordpress.com/2016/12/06/pretty-printing-a-tree-in-a-terminal/)

**Comment by NIFR91** [comment]
Hi, 
My apologies for the late response, somehow I skipped the reviews in my inbox, just recently I checked the pull 
request here on github.  I think I've addressed  all comments please let me know if there's still work to do.

File: contents/tree_traversal/code/crystal/tree-traversal.cr (hunk 6 above)
Comment (jiegillet) [reply contents/tree_traversal/code/crystal/tree-traversal.cr]: Very cool, thanks for the link :)

**Pull request merged**